tests/ProgramTestTests/SimulatedEffects/NavigationTest.elm

```diff
@@ -126,4 +118,45 @@ all =
                 TestingProgram.application SimulatedEffect.Cmd.none
                     |> ProgramTest.expectPageChange "https://example.com/"
                     |> expectFailure [ "expectPageChange: expected to have navigated to a different URL, but no links were clicked and no browser navigation was simulated" ]
+        , pushUrlWithQueryAndFragment
+        ]
+
+
+pushUrlWithQueryAndFragment : Test
+pushUrlWithQueryAndFragment =
+    describe "simulating pushUrl with query and/or fragment and parsing the result gives the correct route" <|
+        let
+            parser =
+                Parser.map
+                    (\q f ->
+                        "NEW"
+                            ++ (Maybe.map ((++) " QUERY:") q |> Maybe.withDefault "")
+                            ++ (Maybe.map ((++) " FRAGMENT:") f |> Maybe.withDefault "")
+                    )
+                    (Parser.s "new" <?> QueryParser.string "query" </> Parser.fragment identity)
+
+            check url expected =
+                test ("pushUrl " ++ url) <|
+                    \_ ->
+                        ProgramTest.createApplication
+                            { onUrlChange = \location -> Log ("Route: " ++ (Parser.parse parser location |> Maybe.withDefault "NOMATCH"))
+                            , onUrlRequest = \_ -> Debug.todo "ProgramTestTests-2:onUrlRequest"
+                            , init = \() location key -> ( [], SimulatedEffect.Cmd.none )
+                            , update = TestingProgram.update
+                            , view =
+                                \_ ->
+                                    { title = "page title"
+                                    , body = []
+                                    }
+                            }
+                            |> ProgramTest.withSimulatedEffects identity
+                            |> ProgramTest.withBaseUrl "https://example.com/path"
+                            |> ProgramTest.start ()
+                            |> ProgramTest.update (ProduceEffects (SimulatedEffect.Navigation.pushUrl url))
+                            |> ProgramTest.expectModel (Expect.equal [ "Route: " ++ expected ])
+        in
+        [ check "new" "NEW"
+        , check "new?query=q" "NEW QUERY:q"
+        , check "new#fragment" "NEW FRAGMENT:fragment"
+        , check "new?query=q#fragment" "NEW QUERY:q FRAGMENT:fragment"
```

Comment:
Sort of. I mainly wrote this test to clearly demonstrate the new issue I mentioned in the PR description.

The difference is that `ExtraTest.elm` currently only compares the stringified versions of the URLs, it doesn't tell you if the URL actually has the query and fragment in the correct place in the URL record and that it would parse correctly into a route.

Reply:
I've now changed the test in `ExtraTest.elm` for "resolve" to compare parsed URLs instead of strings and removed the new test in `NavigationTest.elm`.

Reply:
Okay, that sounds like a good solution